src/Blocks/BlockPlant.h

```diff
@@ -0,0 +1,130 @@
+
+// BlockPlant.h
+
+// Base class for any growing block
+
+
+
+
+
+#pragma once
+
+#include "BlockHandler.h"
+
+class cBlockPlant : public cBlockHandler
+{
+	bool m_NeedLightToGrow;
+public:
+	cBlockPlant(BLOCKTYPE a_BlockType, bool a_LightToGrow)
+		: cBlockHandler(a_BlockType), m_NeedLightToGrow(a_LightToGrow){}
+
+protected:
+	enum PlantAction
+	{
+		paDeath,
+		paGrowth,
+		paStay
+	};
+
+	/// Checks whether there is enough light for the plant to grow.
+	/// If the plant doesn't require light to grow, then it returns paGrowth.
+	/// If the plant requires light to grow and there is enough light, it returns paGrowth.
+	/// If the plant requires light to grow and there isn't enough light, it returns paStay.
+	/// If the plant requires light to grow and there is too little light, it returns paDeath.
+	PlantAction HasEnoughLight(cChunk & a_Chunk, int a_RelX, int a_RelY, int a_RelZ)
+	{
+		// If the plant requires light to grow, check to see if there is enough light
+		// Otherwise, return true
+		if (m_NeedLightToGrow)
+		{
+			NIBBLETYPE Blocklight = a_Chunk.GetBlockLight(a_RelX, a_RelY, a_RelZ);
+			NIBBLETYPE SkyLight = a_Chunk.GetSkyLight  (a_RelX, a_RelY, a_RelZ);
+			NIBBLETYPE Light = a_Chunk.GetTimeAlteredLight(SkyLight);
+
+			// If the amount of light provided by blocks is greater than the sky light, use it instead
+			if (Blocklight > Light)
+			{
+				Light = Blocklight;
+			}
+
+			// Return true if there is enough light
+			// Set m_ShouldDie if the base light amounts are not enough to sustain a plant
+			if (Light > 8)
+			{
+				return paGrowth;
+			}
+			else if (Blocklight < 9 && SkyLight < 9)
+			{
+				return paDeath;
+			}
+
+			return paStay;
+		}
+
+		return paGrowth;
+	}
+
+	/// Checks whether a plant can grow grow, based on what is returned from cBlockPlant::HasEnoughLight
+	/// and a random check based on what is returned from cBlockPlant::GetGrowthChance.
+	/// Can return three values.
+	/// paGrowth when the conditions are right for the plant to grow.
+	/// paStay when the conditions are not quite right.
+	/// paDeath is returned when there isn't enough light for the plant to survive.
+	/// Plants that don't require light will never have a paDeath returned
+	virtual PlantAction CanGrow(cChunk & a_Chunk, int a_RelX, int a_RelY, int a_RelZ)
+	{
+		cFastRandom rand;
+		// Plant can grow if it has the required amount of light, and it passes a random chance based on surrounding blocks
+		PlantAction Action = HasEnoughLight(a_Chunk, a_RelX, a_RelY, a_RelZ);
+		if ((Action == paGrowth) && (rand.NextInt(GetGrowthChance(a_Chunk, a_RelX, a_RelY, a_RelZ)) != 0))
+		{
+			Action = paStay;
+		}
+		return Action;
+	}
+
+	/// Generates a float value based on surrounding blocks that affect how quickly the plant grows.
```

Comment:
Where is the conversion from float to int? GetGrowthChance returns an int.
Range is 0..3.
Int gives me much fewer results, so only a 0%, 50%, 67%, 75% chance of growing.